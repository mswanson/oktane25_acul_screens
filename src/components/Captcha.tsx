import { Control, FieldValues, Path, RegisterOptions } from "react-hook-form";

import { ULThemeFloatingLabelField } from "@/components/form/ULThemeFloatingLabelField";
import { ULThemeFormMessage } from "@/components/form/ULThemeFormMessage";
import { FormField, FormItem } from "@/components/ui/form";
import { cn } from "@/lib/utils";

export interface CaptchaProps<T extends FieldValues = FieldValues> {
  label: string;
  imageUrl: string;
  imageAltText: string;
  name: Path<T>;
  control: Control<T>;
  rules?: RegisterOptions<T>;
  sdkError?: string;
  className?: string;
}

const Captcha = <T extends FieldValues = FieldValues>({
  name,
  control,
  rules,
  label,
  imageUrl,
  imageAltText,
  sdkError,
  className,
}: CaptchaProps<T>) => {
  // Container styles with theme
  const containerStyles = cn("space-y-2", className);

  // Image container styles with theme
  const imageContainerStyles = cn(
    "flex justify-center p-8 rounded",
    "theme-universal:bg-input-bg",
    "theme-universal:border",
    "theme-universal:border-input-border",
    "theme-universal:rounded-input"
  );

  if (!imageUrl) {
    return null;
  }

  return (
    <div className={containerStyles}>
      <div className={imageContainerStyles}>
        <img src={imageUrl} alt={imageAltText} className="object-contain" />
      </div>
      <FormField
        control={control}
        name={name}
        rules={rules}
        render={({ field, fieldState }) => (
          <FormItem>
            <ULThemeFloatingLabelField
              {...field}
              label={label}
              type="text"
              autoComplete="off"
              error={!!fieldState.error || !!sdkError}
            />
            <ULThemeFormMessage
              className="mt-1"
              sdkError={sdkError}
              hasFormError={!!fieldState.error}
            />
          </FormItem>
        )}
      />
    </div>
  );
};

export default Captcha;
