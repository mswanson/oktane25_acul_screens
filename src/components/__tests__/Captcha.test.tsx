import { FormProvider, useForm } from "react-hook-form";

import { render } from "@testing-library/react";

import Captcha from "@/components/Captcha";

// Test wrapper component that provides real form context
function TestWrapper({ children, defaultValues = {} }: any) {
  const methods = useForm({ defaultValues });
  return (
    <FormProvider {...methods}>
      <form>{children({ control: methods.control })}</form>
    </FormProvider>
  );
}

describe("Captcha Component", () => {
  const baseProps = {
    label: "Enter CAPTCHA",
    imageUrl: "https://example.com/captcha.png",
    imageAltText: "CAPTCHA verification image",
    name: "captcha" as const,
  };

  // Snapshot Tests
  it("matches snapshot when imageUrl is provided", () => {
    const { container } = render(
      <TestWrapper>
        {({ control }: any) => <Captcha {...baseProps} control={control} />}
      </TestWrapper>
    );
    expect(container.firstChild).toMatchSnapshot();
  });

  it("matches snapshot when imageUrl is empty", () => {
    const { container } = render(
      <TestWrapper>
        {({ control }: any) => (
          <Captcha {...baseProps} imageUrl="" control={control} />
        )}
      </TestWrapper>
    );
    expect(container.firstChild).toMatchSnapshot();
  });

  it("matches snapshot with custom className", () => {
    const { container } = render(
      <TestWrapper>
        {({ control }: any) => (
          <Captcha {...baseProps} className="custom-class" control={control} />
        )}
      </TestWrapper>
    );
    expect(container.firstChild).toMatchSnapshot();
  });

  // Functional Tests
  it("renders CAPTCHA image when imageUrl is provided", () => {
    const { container } = render(
      <TestWrapper>
        {({ control }: any) => <Captcha {...baseProps} control={control} />}
      </TestWrapper>
    );

    const image = container.querySelector("img");
    expect(image).toBeInTheDocument();
    expect(image).toHaveAttribute("src", "https://example.com/captcha.png");
    expect(image).toHaveAttribute("alt", "CAPTCHA verification image");
  });

  it("does not render when imageUrl is empty", () => {
    const { container } = render(
      <TestWrapper>
        {({ control }: any) => (
          <Captcha {...baseProps} imageUrl="" control={control} />
        )}
      </TestWrapper>
    );

    expect(container.firstChild?.firstChild).toBeNull();
  });

  it("applies theme classes to image container", () => {
    const { container } = render(
      <TestWrapper>
        {({ control }: any) => <Captcha {...baseProps} control={control} />}
      </TestWrapper>
    );

    // Find the image container div that contains the img element
    const imageContainer = container.querySelector("div img")?.parentElement;
    expect(imageContainer).toHaveClass("theme-universal:bg-input-bg");
    expect(imageContainer).toHaveClass("theme-universal:border-input-border");
    expect(imageContainer).toHaveClass("theme-universal:rounded-input");
  });

  it("applies custom className when provided", () => {
    const { container } = render(
      <TestWrapper>
        {({ control }: any) => (
          <Captcha {...baseProps} className="custom-class" control={control} />
        )}
      </TestWrapper>
    );

    const rootContainer = container.querySelector(".custom-class");
    expect(rootContainer).toBeInTheDocument();
  });
});
