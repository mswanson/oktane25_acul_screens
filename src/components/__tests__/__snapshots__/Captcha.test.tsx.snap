// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`Captcha Component matches snapshot when imageUrl is empty 1`] = `<form />`;

exports[`Captcha Component matches snapshot when imageUrl is provided 1`] = `
<form>
  <div
    class="space-y-2"
  >
    <div
      class="flex justify-center p-8 rounded theme-universal:bg-input-bg theme-universal:border theme-universal:border-input-border theme-universal:rounded-input"
    >
      <img
        alt="CAPTCHA verification image"
        class="object-contain"
        src="https://example.com/captcha.png"
      />
    </div>
    <div
      class="grid"
      data-slot="form-item"
    >
      <div
        class="w-full"
      >
        <div>
          <div
            class="bg-input aria-invalid:ring-destructive/20 dark:aria-invalid:ring-destructive/40 aria-invalid:border-destructive theme-default:active:scale-[0.99] box-border inline-flex w-full shrink-0 cursor-text justify-center text-sm transition-[color,box-shadow] duration-150 ease-in-out outline-none disabled:pointer-events-none disabled:opacity-50 [&_svg]:pointer-events-none [&_svg]:shrink-0 [&_svg:not([class*='size-'])]:size-4 text-input-foreground shadow-input-resting hover:shadow-input-hover focus-within:ring-ring focus-within:ring-4 h-14 rounded-2xl group relative items-end gap-0.5 mb-2 theme-universal:bg-input-bg theme-universal:text-input-text theme-universal:border-(length:--ul-theme-border-input-border-weight) theme-universal:border-input-border theme-universal:rounded-input theme-universal:text-(length:--ul-theme-font-body-text-size) theme-universal:font-body-text theme-universal:placeholder:text-input-labels theme-universal:[&_input]:placeholder:text-input-labels theme-universal:placeholder:text-(length:--ul-theme-font-input-labels-size) theme-universal:[&_input]:placeholder:text-(length:--ul-theme-font-input-labels-size) theme-universal:placeholder:font-input-label theme-universal:[&_input]:placeholder:font-input-label theme-universal:[&_label]:bg-input-bg theme-universal:[&_label]:text-input-labels theme-universal:[&_label]:text-(length:--ul-theme-font-input-labels-size) theme-universal:[&_label]:font-input-label theme-universal:focus-within:border-base-focus theme-universal:focus-within:ring-0 theme-universal:focus-within:[&_label]:text-base-focus"
          >
            <label
              class="pointer-events-none absolute top-1/2 left-4 z-[10] -translate-y-1/2 text-sm transition-all duration-150 ease-in-out origin-left text-muted-foreground"
              for="«r0»-form-item"
              id="base-ui-«r1»"
            >
              Enter CAPTCHA
            </label>
            <div
              class="relative flex-1"
            >
              <input
                aria-labelledby="base-ui-«r1»"
                autocomplete="off"
                class="h-14 w-full flex-1 rounded-2xl bg-transparent px-3 py-4 outline-none file:border-0 file:bg-transparent file:text-sm file:font-medium autofill:bg-transparent autofill:[-webkit-text-fill-color:inherit] autofill:[transition:background-color_9999s_ease-in-out_0s] pl-3 pr-3"
                id="«r0»-form-item"
                name="captcha"
                type="text"
              />
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</form>
`;

exports[`Captcha Component matches snapshot with custom className 1`] = `
<form>
  <div
    class="space-y-2 custom-class"
  >
    <div
      class="flex justify-center p-8 rounded theme-universal:bg-input-bg theme-universal:border theme-universal:border-input-border theme-universal:rounded-input"
    >
      <img
        alt="CAPTCHA verification image"
        class="object-contain"
        src="https://example.com/captcha.png"
      />
    </div>
    <div
      class="grid"
      data-slot="form-item"
    >
      <div
        class="w-full"
      >
        <div>
          <div
            class="bg-input aria-invalid:ring-destructive/20 dark:aria-invalid:ring-destructive/40 aria-invalid:border-destructive theme-default:active:scale-[0.99] box-border inline-flex w-full shrink-0 cursor-text justify-center text-sm transition-[color,box-shadow] duration-150 ease-in-out outline-none disabled:pointer-events-none disabled:opacity-50 [&_svg]:pointer-events-none [&_svg]:shrink-0 [&_svg:not([class*='size-'])]:size-4 text-input-foreground shadow-input-resting hover:shadow-input-hover focus-within:ring-ring focus-within:ring-4 h-14 rounded-2xl group relative items-end gap-0.5 mb-2 theme-universal:bg-input-bg theme-universal:text-input-text theme-universal:border-(length:--ul-theme-border-input-border-weight) theme-universal:border-input-border theme-universal:rounded-input theme-universal:text-(length:--ul-theme-font-body-text-size) theme-universal:font-body-text theme-universal:placeholder:text-input-labels theme-universal:[&_input]:placeholder:text-input-labels theme-universal:placeholder:text-(length:--ul-theme-font-input-labels-size) theme-universal:[&_input]:placeholder:text-(length:--ul-theme-font-input-labels-size) theme-universal:placeholder:font-input-label theme-universal:[&_input]:placeholder:font-input-label theme-universal:[&_label]:bg-input-bg theme-universal:[&_label]:text-input-labels theme-universal:[&_label]:text-(length:--ul-theme-font-input-labels-size) theme-universal:[&_label]:font-input-label theme-universal:focus-within:border-base-focus theme-universal:focus-within:ring-0 theme-universal:focus-within:[&_label]:text-base-focus"
          >
            <label
              class="pointer-events-none absolute top-1/2 left-4 z-[10] -translate-y-1/2 text-sm transition-all duration-150 ease-in-out origin-left text-muted-foreground"
              for="«r3»-form-item"
              id="base-ui-«r4»"
            >
              Enter CAPTCHA
            </label>
            <div
              class="relative flex-1"
            >
              <input
                aria-labelledby="base-ui-«r4»"
                autocomplete="off"
                class="h-14 w-full flex-1 rounded-2xl bg-transparent px-3 py-4 outline-none file:border-0 file:bg-transparent file:text-sm file:font-medium autofill:bg-transparent autofill:[-webkit-text-fill-color:inherit] autofill:[transition:background-color_9999s_ease-in-out_0s] pl-3 pr-3"
                id="«r3»-form-item"
                name="captcha"
                type="text"
              />
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</form>
`;
